datadog/dogstatsd/base.py

```diff
@@ -362,12 +374,25 @@ def open_buffer(self, max_buffer_size=None):
     def close_buffer(self):
         """
         Flush the buffer and switch back to single metric packets.
+
+        Note: This method must be called after a matching open_buffer()
+        invocation.
         """
-        self._send = self._send_to_server
 
-        if self.buffer:
-            # Only send packets if there are packets to send
-            self._flush_buffer()
+        try:
+            if not hasattr(self, 'buffer'):
+                raise BufferError('Cannot close buffer that was never opened')
+
+            self._send = self._send_to_server
+
+            if self.buffer:
+                # Only send packets if there are packets to send
+                self._flush_buffer()
+        finally:
+            # Release all locks that might have been held by our thread.
+            while self._buffer_lock_depth > 0:
```

Comment:
Consider the following code:
```python
d = DogStatsd()
# ...
with d as batch:
       with d as other_batch:
             other_batch.gauge("users.online", 123)
       # As the lock was released at the previous line, another thread can call `d.open_buffer()` and update buffer at the same time as `batch.gauge("users.online", 123)`
       batch.gauge("users.online", 123)
```

> As a sidenote, both open_buffer and close_buffer are getting deprecated soon since we want buffering by default so they will effectively be noops in the future

Note: I am not sure it worth handling the case in the previous code as these functions are getting deprecated.

Reply:
That's an interesting edge case for sure! Let me think about this some more when I get time and I'll get back to you on it.

Reply:
After some thinking about this particular example and other edge use cases, I think it might be just better to make the lock not re-entrant from the same thread which should fix your and @jd's concerns.